File: Python/589. N-ary Tree Preorder Traversal.py
from typing import List

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        ret, q = [], root and [root]
        while q:
            node = q.pop()
            ret.append(node.val)
            q += [child for child in reversed(node.children) if child]
        return ret

File: JavaScript/589. N-ary Tree Preorder Traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node|null} root
 * @return {number[]}
 */

const preorder = (root) => {
    const ret = []
    const q = [root]

    while (q.length) {
        const node = q.pop()
        if (node) {
            const {children, val} = node
            ret.push(val)
            if (children) {
                children.reverse()
                children.forEach(node => {
                    if (node) q.push(node)
                })
            }
        }
    }
    return ret
}